File: routes/lorebooks.py
import json
from flask import Blueprint, request, jsonify
from db import get_db

lorebooks_bp = Blueprint('lorebooks', __name__)


def _entry_to_dict(row):
    d = dict(row)
    for field in ('keys', 'secondary_keys'):
        raw = d.get(field)
        try:
            d[field] = json.loads(raw) if raw else ([] if field == 'keys' else None)
        except (json.JSONDecodeError, TypeError):
            d[field] = [] if field == 'keys' else None
    return d


@lorebooks_bp.route('/get_lorebooks', methods=['GET'])
def get_lorebooks():
    """Весь список лорбуков юзера — вкладка Лорбуков. Не фильтрует по
    shared/owner здесь: это полный список, фронт решает что показывать
    (например 'мои' vs 'общедоступные' как табы над одним и тем же списком)."""
    try:
        db = get_db()
        rows = db.execute('SELECT * FROM lorebooks ORDER BY id').fetchall()
        return jsonify([dict(r) for r in rows])
    except Exception as e:
        return jsonify({'error': str(e)}), 500


@lorebooks_bp.route('/lorebook/<int:lorebook_id>', methods=['GET'])
def get_lorebook(lorebook_id):
    """Один лорбук + все его entries — для экрана редактирования."""
    try:
        db = get_db()
        lorebook = db.execute('SELECT * FROM lorebooks WHERE id = ?', (lorebook_id,)).fetchone()
        if not lorebook:
            return jsonify({'error': 'lorebook not found'}), 404
        entries = db.execute(
            'SELECT * FROM lorebook_entries WHERE lorebook_id = ? ORDER BY priority DESC, id',
            (lorebook_id,)
        ).fetchall()
        result = dict(lorebook)
        result['entries'] = [_entry_to_dict(e) for e in entries]
        return jsonify(result)
    except Exception as e:
        return jsonify({'error': str(e)}), 500


@lorebooks_bp.route('/save_lorebook', methods=['POST'])
def save_lorebook():
    """Создаёт лорбук (пустой или с entries сразу, если фронт шлёт их
    массивом — удобно для импорта из карточки персонажа, см. отдельный
    /import_lorebook_from_character ниже, который переиспользует это же
    поле). auto_created_for_character_id не принимается отсюда напрямую —
    для этого есть отдельный явный эндпоинт импорта, чтобы обычное ручное
    'создать лорбук' не могло случайно выставить происхождение."""
    try:
        data = request.json or {}
        name = data.get('name', 'New lorebook')
        is_shared = 1 if data.get('is_shared') else 0
        scan_depth = data.get('scan_depth', 4)
        entries = data.get('entries', [])

        db = get_db()
        cur = db.execute(
            'INSERT INTO lorebooks (name, is_shared, scan_depth) VALUES (?, ?, ?)',
            (name, is_shared, scan_depth)
        )
        lorebook_id = cur.lastrowid
        _insert_entries(db, lorebook_id, entries)
        db.commit()
        return jsonify({'success': True, 'id': lorebook_id})
    except Exception as e:
        return jsonify({'success': False, 'error': str(e)}), 500


def _insert_entries(db, lorebook_id, entries):
    for e in entries:
        db.execute(
            '''INSERT INTO lorebook_entries (
                   lorebook_id, keys, secondary_keys, content, enabled,
                   priority, case_sensitive, position, token_budget
               ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)''',
            (
                lorebook_id,
                json.dumps(e.get('keys', []), ensure_ascii=False),
                json.dumps(e.get('secondary_keys'), ensure_ascii=False) if e.get('secondary_keys') else None,
                e.get('content', ''),
                1 if e.get('enabled', True) else 0,
                e.get('priority', 100),
                1 if e.get('case_sensitive') else 0,
                e.get('position', 'before_char'),
                e.get('token_budget'),
            )
        )


@lorebooks_bp.route('/update_lorebook/<int:lorebook_id>', methods=['PUT'])
def update_lorebook(lorebook_id):
    """Обновляет мета-поля лорбука (name/is_shared/scan_depth) И полностью
    заменяет набор entries, если он передан — тот же 'снести и перезаписать'
    подход что в save_chat_history, оправдан по той же причине: фронт
    редактирует entries как единый список в форме, дельту присылать не
    удобнее чем целиком."""
    try:
        data = request.json or {}
        db = get_db()
        existing = db.execute('SELECT id FROM lorebooks WHERE id = ?', (lorebook_id,)).fetchone()
        if not existing:
            return jsonify({'success': False, 'error': 'lorebook not found'}), 404

        db.execute(
            '''UPDATE lorebooks SET name = ?, is_shared = ?, scan_depth = ?,
                   updated_at = datetime('now') WHERE id = ?''',
            (
                data.get('name', 'Unnamed'),
                1 if data.get('is_shared') else 0,
                data.get('scan_depth', 4),
                lorebook_id,
            )
        )

        if 'entries' in data:
            db.execute('DELETE FROM lorebook_entries WHERE lorebook_id = ?', (lorebook_id,))
            _insert_entries(db, lorebook_id, data['entries'])

        db.commit()
        return jsonify({'success': True})
    except Exception as e:
        return jsonify({'success': False, 'error': str(e)}), 500


@lorebooks_bp.route('/rename_lorebook/<int:lorebook_id>', methods=['PUT'])
def rename_lorebook(lorebook_id):
    """Отдельный узкий эндпоинт для переименования (обсуждали отдельно от
    остального редактирования) — не требует гонять entries туда-обратно
    ради смены одного поля."""
    try:
        data = request.json or {}
        name = data.get('name', '').strip()
        if not name:
            return jsonify({'success': False, 'error': 'name required'}), 400
        db = get_db()
        cur = db.execute(
            "UPDATE lorebooks SET name = ?, updated_at = datetime('now') WHERE id = ?",
            (name, lorebook_id)
        )
        db.commit()
        if cur.rowcount == 0:
            return jsonify({'success': False, 'error': 'lorebook not found'}), 404
        return jsonify({'success': True})
    except Exception as e:
        return jsonify({'success': False, 'error': str(e)}), 500


@lorebooks_bp.route('/delete_lorebook/<int:lorebook_id>', methods=['DELETE'])
def delete_lorebook(lorebook_id):
    """Просто удаляет — CASCADE сносит entries и character_lorebooks сам.
    Предупреждение 'это был локальный лорбук, точно удалить?' — та
    развилка, которую обсуждали ('автосоздан под перса и больше никуда не
    подключен') — это чисто UI-уровень принятия решения ПЕРЕД вызовом
    этого эндпоинта: фронт сам смотрит is_shared + сколько персонажей
    подключено (через /lorebook/<id>/characters ниже) и решает, показывать
    ли предупреждение. Здесь на бэкенде разделять 'просто отвязать' и
    'удалить насовсем' незачем — раз юзер дошёл до этого DELETE, значит
    уже решил, оба сценария ('локальный, никуда не подключен' и 'общий,
    подключен много где') одинаково завершаются полным удалением записи."""
    try:
        db = get_db()
        db.execute('DELETE FROM lorebooks WHERE id = ?', (lorebook_id,))
        db.commit()
        return jsonify({'success': True})
    except Exception as e:
        return jsonify({'success': False, 'error': str(e)}), 500


# --- Подключение к персонажам (N:N через character_lorebooks) -------------

@lorebooks_bp.route('/character/<int:character_id>/lorebooks', methods=['GET'])
def get_character_lorebooks(character_id):
    try:
        db = get_db()
        rows = db.execute(
            '''SELECT l.* FROM lorebooks l
               JOIN character_lorebooks cl ON cl.lorebook_id = l.id
               WHERE cl.character_id = ?
               ORDER BY l.id''',
            (character_id,)
        ).fetchall()
        return jsonify([dict(r) for r in rows])
    except Exception as e:
        return jsonify({'error': str(e)}), 500


@lorebooks_bp.route('/lorebook/<int:lorebook_id>/characters', methods=['GET'])
def get_lorebook_characters(lorebook_id):
    """Нужно фронту чтобы посчитать 'сколько персонажей подключено' перед
    показом предупреждения при удалении (см. комментарий в delete_lorebook)."""
    try:
        db = get_db()
        rows = db.execute(
            '''SELECT c.id, c.name FROM characters c
               JOIN character_lorebooks cl ON cl.character_id = c.id
               WHERE cl.lorebook_id = ?
               ORDER BY c.id''',
            (lorebook_id,)
        ).fetchall()
        return jsonify([dict(r) for r in rows])
    except Exception as e:
        return jsonify({'error': str(e)}), 500


@lorebooks_bp.route('/connect_lorebook', methods=['POST'])
def connect_lorebook():
    try:
        data = request.json or {}
        character_id = data.get('character_id')
        lorebook_id = data.get('lorebook_id')
        if not character_id or not lorebook_id:
            return jsonify({'success': False, 'error': 'character_id and lorebook_id required'}), 400
        db = get_db()
        db.execute(
            'INSERT OR IGNORE INTO character_lorebooks (character_id, lorebook_id) VALUES (?, ?)',
            (character_id, lorebook_id)
        )
        db.commit()
        return jsonify({'success': True})
    except Exception as e:
        return jsonify({'success': False, 'error': str(e)}), 500


@lorebooks_bp.route('/disconnect_lorebook', methods=['POST'])
def disconnect_lorebook():
    try:
        data = request.json or {}
        character_id = data.get('character_id')
        lorebook_id = data.get('lorebook_id')
        if not character_id or not lorebook_id:
            return jsonify({'success': False, 'error': 'character_id and lorebook_id required'}), 400
        db = get_db()
        db.execute(
            'DELETE FROM character_lorebooks WHERE character_id = ? AND lorebook_id = ?',
            (character_id, lorebook_id)
        )
        db.commit()
        return jsonify({'success': True})
    except Exception as e:
        return jsonify({'success': False, 'error': str(e)}), 500


@lorebooks_bp.route('/import_lorebook_from_character', methods=['POST'])
def import_lorebook_from_character():
    """Автосоздание лорбука из character_book при импорте PNG-карточки
    (см. routes/characters.py: import_character_png теперь отдаёт
    raw_character_book отдельным полем вместо того чтобы тихо его
    дропать — фронт после успешного импорта персонажа зовёт этот
    эндпоинт, если raw_character_book.entries непустой).

    Ожидает: {character_id, name, entries: [...]}
    Создаёт лорбук с auto_created_for_character_id = character_id и сразу
    подключает его к этому персонажу через character_lorebooks — ровно то
    'автоматически создаётся и подключается', что обсуждали. Возможность
    докинуть ещё общих лорбуков сверху не отменяется — это просто ещё один
    connect_lorebook на тот же character_id."""
    try:
        data = request.json or {}
        character_id = data.get('character_id')
        name = data.get('name', 'Imported lorebook')
        entries = data.get('entries', [])
        if not character_id:
            return jsonify({'success': False, 'error': 'character_id required'}), 400
        if not entries:
            return jsonify({'success': False, 'error': 'no entries to import'}), 400

        db = get_db()
        char = db.execute('SELECT id FROM characters WHERE id = ?', (character_id,)).fetchone()
        if not char:
            return jsonify({'success': False, 'error': 'character not found'}), 404

        cur = db.execute(
            'INSERT INTO lorebooks (name, is_shared, auto_created_for_character_id) VALUES (?, 0, ?)',
            (name, character_id)
        )
        lorebook_id = cur.lastrowid
        _insert_entries(db, lorebook_id, entries)
        db.execute(
            'INSERT OR IGNORE INTO character_lorebooks (character_id, lorebook_id) VALUES (?, ?)',
            (character_id, lorebook_id)
        )
        db.commit()
        return jsonify({'success': True, 'lorebook_id': lorebook_id, 'imported_entries': len(entries)})
    except Exception as e:
        return jsonify({'success': False, 'error': str(e)}), 500

File: static/js/lorebooks.js
// static/js/lorebooks.js
// Список лорбуков + редактор (модалка) с entries. Новый файл — заменяет
// собой то, что раньше было character_book внутри карточки персонажа.

let _lorebooksCache = [];
let editingLorebookId = null;
let _lorebookEntryCounter = 0;

async function fetchLorebooks() {
    const r = await fetch(`${BASE_URL}/get_lorebooks`);
    return r.json();
}

async function fetchLorebook(id) {
    const r = await fetch(`${BASE_URL}/lorebook/${id}`);
    return r.json();
}

function loadLorebookList() {
    const container = document.getElementById('lorebookGrid');
    if (!container) return;
    container.innerHTML = '<div class="grid-full-center"><div class="loader-spinner"></div></div>';
    fetchLorebooks().then(lorebooks => {
        _lorebooksCache = lorebooks;
        renderLorebookCards(lorebooks);
    }).catch(e => {
        console.error('Error loading lorebooks:', e);
        container.innerHTML = '<div class="empty-state">Failed to load</div>';
    });
}

function renderLorebookCards(lorebooks) {
    const container = document.getElementById('lorebookGrid');
    if (!container) return;
    if (!lorebooks.length) {
        container.innerHTML = '<div class="empty-state">No lorebooks yet</div>';
        return;
    }
    container.innerHTML = lorebooks.map(lb => `
        <div class="character-card compact" data-lorebook-id="${lb.id}">
            <div class="character-card-clickable" onclick="openLorebookEditor(${lb.id})">
                <div class="character-card-name">
                    ${lb.name}
                    ${lb.is_shared ? '<span class="field-tag-muted">shared</span>' : ''}
                </div>
                <div class="character-card-desc">
                    ${lb.auto_created_for_character_id ? 'Auto-created from a character card' : 'Manually created'}
                </div>
            </div>
            <div class="character-actions">
                <button class="action-btn delete-btn" onclick="deleteLorebook(${lb.id}, event)">
                    <img src="/static/icons/trash.svg" alt="delete">
                </button>
            </div>
        </div>
    `).join('');
    if (typeof _staggerCardReveal === 'function') _staggerCardReveal(container);
}

function resetLorebookEditor() {
    editingLorebookId = null;
    document.getElementById('lorebookEditorTitle').textContent = 'New lorebook';
    document.getElementById('lorebookName').value = '';
    document.getElementById('lorebookShared').checked = false;
    document.getElementById('lorebookScanDepth').value = 4;
    document.getElementById('lorebookEntriesList').innerHTML = '';
    _updateLorebookEntriesHeader();
}

async function openLorebookEditor(id = null) {
    const modal = document.getElementById('lorebookEditorModal');
    if (!modal) return;

    if (id) {
        editingLorebookId = id;
        document.getElementById('lorebookEditorTitle').textContent = 'Edit lorebook';
        try {
            const lb = await fetchLorebook(id);
            document.getElementById('lorebookName').value = lb.name || '';
            document.getElementById('lorebookShared').checked = !!lb.is_shared;
            document.getElementById('lorebookScanDepth').value = lb.scan_depth || 4;
            const list = document.getElementById('lorebookEntriesList');
            list.innerHTML = '';
            (lb.entries || []).forEach(entry => addLorebookEntryItem(entry));
            _updateLorebookEntriesHeader();
        } catch (e) {
            showCustomAlert('Failed to load lorebook: ' + e);
            return;
        }
    } else {
        resetLorebookEditor();
    }

    modal.classList.add('active');
}

function closeLorebookEditor() {
    _closeModalAnimated(document.getElementById('lorebookEditorModal'));
}

// Каждая entry — свой мини-блок: keys (comma-separated), content, priority,
// enabled. secondary_keys/case_sensitive/position/token_budget не выведены
// в UI этой грубой версии (доступны через API, просто не редактируются
// отсюда) — хранятся as-is при round-trip если были заданы раньше (через
// data-атрибуты, не теряются молча).
function addLorebookEntryItem(entry = {}) {
    const list = document.getElementById('lorebookEntriesList');
    if (!list) return;
    const uid = `lb-entry-${_lorebookEntryCounter++}`;

    const item = document.createElement('div');
    item.className = 'lorebook-entry-item field-accordion open';
    item.dataset.entryUid = uid;
    item.dataset.secondaryKeys = entry.secondary_keys ? JSON.stringify(entry.secondary_keys) : '';
    item.dataset.caseSensitive = entry.case_sensitive ? '1' : '0';
    item.dataset.position = entry.position || 'before_char';
    item.dataset.tokenBudget = entry.token_budget != null ? entry.token_budget : '';

    item.innerHTML = `
        <div class="field-accordion-header" onclick="this.parentElement.classList.toggle('open')">
            <span class="field-accordion-title">${(entry.keys || []).join(', ') || 'New entry'}</span>
            <button class="field-remove-btn" onclick="event.stopPropagation(); this.closest('.lorebook-entry-item').remove(); _updateLorebookEntriesHeader()">×</button>
        </div>
        <div class="field-accordion-body">
            <div class="field-accordion-body-inner">
                <label class="field-label-sm">Keys (comma-separated)</label>
                <input type="text" class="field-input lorebook-entry-keys" value="${(entry.keys || []).join(', ')}" placeholder="dragon, wyrm">
                <label class="field-label-sm">Content</label>
                <textarea class="field-textarea lorebook-entry-content" placeholder="What gets injected when a key matches…">${entry.content || ''}</textarea>
                <label class="field-label-sm">Priority <span class="field-tag-muted">higher wins the token budget first</span></label>
                <input type="number" class="field-input lorebook-entry-priority" value="${entry.priority ?? 100}">
                <label class="field-checkbox-label">
                    <input type="checkbox" class="lorebook-entry-enabled" ${entry.enabled === false ? '' : 'checked'}>
                    <span>Enabled</span>
                </label>
            </div>
        </div>`;
    list.appendChild(item);
    _updateLorebookEntriesHeader();
}

function _updateLorebookEntriesHeader() {
    const count = document.querySelectorAll('#lorebookEntriesList .lorebook-entry-item').length;
    const label = document.getElementById('lorebookEntriesLabel');
    if (label) label.textContent = `Entries (${count})`;
}

function _collectLorebookEntries() {
    return Array.from(document.querySelectorAll('#lorebookEntriesList .lorebook-entry-item')).map(item => {
        const keys = item.querySelector('.lorebook-entry-keys').value
            .split(',').map(k => k.trim()).filter(Boolean);
        const content = item.querySelector('.lorebook-entry-content').value.trim();
        const priority = parseInt(item.querySelector('.lorebook-entry-priority').value, 10) || 100;
        const enabled = item.querySelector('.lorebook-entry-enabled').checked;

        const secondaryKeysRaw = item.dataset.secondaryKeys;
        let secondary_keys = null;
        if (secondaryKeysRaw) {
            try { secondary_keys = JSON.parse(secondaryKeysRaw); } catch (e) { secondary_keys = null; }
        }
        const tokenBudgetRaw = item.dataset.tokenBudget;

        return {
            keys, content, priority, enabled,
            secondary_keys,
            case_sensitive: item.dataset.caseSensitive === '1',
            position: item.dataset.position || 'before_char',
            token_budget: tokenBudgetRaw ? parseInt(tokenBudgetRaw, 10) : null,
        };
    });
}

function saveLorebook() {
    const name = document.getElementById('lorebookName').value.trim();
    if (!name) { showCustomAlert('Enter a lorebook name'); return; }

    const payload = {
        name,
        is_shared: document.getElementById('lorebookShared').checked,
        scan_depth: parseInt(document.getElementById('lorebookScanDepth').value, 10) || 4,
        entries: _collectLorebookEntries(),
    };

    const url = editingLorebookId
        ? `${BASE_URL}/update_lorebook/${editingLorebookId}`
        : `${BASE_URL}/save_lorebook`;
    const method = editingLorebookId ? 'PUT' : 'POST';

    fetch(url, {
        method,
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(payload)
    }).then(r => r.json()).then(data => {
        if (data.success) {
            closeLorebookEditor();
            loadLorebookList();
        } else {
            showCustomAlert('Save error: ' + (data.error || 'unknown'));
        }
    }).catch(e => showCustomAlert('Error: ' + e));
}

// Удаление: спрашиваем то же "локальный неподключённый лорбук?" на
// фронте, раз бэкенд сознательно не делает эту проверку сам (см.
// routes/lorebooks.py comment — UI-уровня решение). Считаем подключения
// перед удалением и предупреждаем только если лорбук реально к чему-то
// привязан ИЛИ помечен shared — иначе тихое confirm без лишних деталей.
function deleteLorebook(id, event) {
    event.stopPropagation();
    fetch(`${BASE_URL}/lorebook/${id}/characters`).then(r => r.json()).then(characters => {
        const lb = _lorebooksCache.find(l => l.id === id);
        const usedElsewhere = (characters || []).length > 0;
        const isShared = lb?.is_shared;

        let message = 'Delete this lorebook?';
        if (usedElsewhere) {
            message = `This lorebook is connected to ${characters.length} character(s). Deleting it removes it from all of them. Continue?`;
        } else if (isShared) {
            message = 'This lorebook is marked shared but not connected to any character right now. Delete it?';
        }

        showConfirm(message, () => {
            fetch(`${BASE_URL}/delete_lorebook/${id}`, { method: 'DELETE' })
                .then(r => r.json()).then(data => {
                    if (data.success) loadLorebookList();
                    else showCustomAlert('Delete error');
                }).catch(e => showCustomAlert('Error: ' + e));
        });
    }).catch(() => {
        showConfirm('Delete this lorebook?', () => {
            fetch(`${BASE_URL}/delete_lorebook/${id}`, { method: 'DELETE' })
                .then(r => r.json()).then(data => {
                    if (data.success) loadLorebookList();
                    else showCustomAlert('Delete error');
                }).catch(e => showCustomAlert('Error: ' + e));
        });
    });
}

// ── Подключение лорбуков к персонажу ────────────────────────────────
// Симметрично connected-personas на характерс-стороне — рендерится в
// новый аккордеон в форме персонажа (см. _forms.html правку ниже).

async function fetchCharacterLorebooks(characterId) {
    const r = await fetch(`${BASE_URL}/character/${characterId}/lorebooks`);
    return r.json();
}

async function renderCharacterConnectedLorebooks(characterId) {
    const container = document.getElementById('characterConnectedLorebooks');
    const select = document.getElementById('connectLorebookSelect');
    if (!container) return;

    if (!characterId) {
        container.innerHTML = '';
        if (select) select.innerHTML = '';
        return;
    }

    container.innerHTML = '<div class="loader-spinner"></div>';
    try {
        const [connected, allLorebooks] = await Promise.all([
            fetchCharacterLorebooks(characterId),
            fetchLorebooks()
        ]);

        if (!connected.length) {
            container.innerHTML = '<div class="empty-state-small">No lorebooks connected yet</div>';
        } else {
            container.innerHTML = connected.map(lb => `
                <div class="connected-chip" data-lorebook-id="${lb.id}">
                    <span>${lb.name}</span>
                    <button class="chip-remove-btn" onclick="disconnectCharacterFromLorebook(${characterId}, ${lb.id})">×</button>
                </div>
            `).join('');
        }

        if (select) {
            const connectedIds = new Set(connected.map(lb => lb.id));
            const selectable = allLorebooks.filter(lb => !connectedIds.has(lb.id));
            select.innerHTML = selectable.length
                ? selectable.map(lb => `<option value="${lb.id}">${lb.name}${lb.is_shared ? ' (shared)' : ''}</option>`).join('')
                : '<option value="" disabled selected>No more lorebooks to connect</option>';
        }
    } catch (e) {
        console.error('Error loading connected lorebooks:', e);
        container.innerHTML = '';
    }
}

function disconnectCharacterFromLorebook(characterId, lorebookId) {
    fetch(`${BASE_URL}/disconnect_lorebook`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ character_id: characterId, lorebook_id: lorebookId })
    }).then(r => r.json()).then(data => {
        if (data.success) renderCharacterConnectedLorebooks(characterId);
        else showCustomAlert('Disconnect error');
    }).catch(e => showCustomAlert('Error: ' + e));
}

function connectLorebookToCurrentCharacter() {
    if (!editingCharacterId) {
        showCustomAlert('Save the character first, then connect lorebooks to it.');
        return;
    }
    const select = document.getElementById('connectLorebookSelect');
    const lorebookId = select ? parseInt(select.value, 10) : null;
    if (!lorebookId) return;

    fetch(`${BASE_URL}/connect_lorebook`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ character_id: editingCharacterId, lorebook_id: lorebookId })
    }).then(r => r.json()).then(data => {
        if (data.success) renderCharacterConnectedLorebooks(editingCharacterId);
        else showCustomAlert('Connect error');
    }).catch(e => showCustomAlert('Error: ' + e));
}
